File: android/src/main/java/cn/reactnative/modules/update/UpdateContext.java
package cn.reactnative.modules.update;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.util.Log;
import com.facebook.react.ReactInstanceManager;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import java.io.File;
import java.net.URL;

/**
 * Created by tdzl2003 on 3/31/16.
 */
public class UpdateContext {
    private Context context;
    private File rootDir;
    private Executor executor;

    public static boolean DEBUG = false;
    private static ReactInstanceManager mReactInstanceManager;
    private static boolean isUsingBundleUrl = false;

    public UpdateContext(Context context) {
        this.context = context;
        this.executor = Executors.newSingleThreadExecutor();

        this.rootDir = new File(context.getFilesDir(), "_update");

        if (!rootDir.exists()) {
            rootDir.mkdir();
        }

        this.sp = context.getSharedPreferences("update", Context.MODE_PRIVATE);

        String packageVersion = getPackageVersion();
        if (!packageVersion.equals(this.sp.getString("packageVersion", null))) {
            SharedPreferences.Editor editor = sp.edit();
            editor.clear();
            editor.putString("packageVersion", packageVersion);
            editor.apply();

            this.cleanUp();
        }
    }

    public String getRootDir() {
        return rootDir.toString();
    }

    public String getPackageVersion() {
        PackageManager pm = context.getPackageManager();
        PackageInfo pi = null;
        try {
            pi = pm.getPackageInfo(context.getPackageName(), 0);
            return pi.versionName;
        } catch( PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        return null;
    }

    public String getBuildTime() {
        return context.getString(R.string.pushy_build_time);
    }

    public String getUuid() {
        return sp.getString("uuid", null);
    }

    public Map getBlockUpdate() {
        return new HashMap<String, Object>() {{
            put("until", sp.getInt("blockUntil", 0));
            put("reason", sp.getString("blockReason", null));
        }};
    }

    public boolean getIsUsingBundleUrl() {
        return isUsingBundleUrl;
    }

    public interface DownloadFileListener {
        void onDownloadCompleted(DownloadTaskParams params);
        void onDownloadFailed(Throwable error);
    }

    public void downloadFullUpdate(String url, String hash, DownloadFileListener listener) {
        DownloadTaskParams params = new DownloadTaskParams();
        params.type = DownloadTaskParams.TASK_TYPE_PATCH_FULL;
        params.url = url;
        params.hash = hash;
        params.listener = listener;
        params.targetFile = new File(rootDir, hash + ".ppk");
        params.unzipDirectory = new File(rootDir, hash);
        new DownloadTask(context).executeOnExecutor(this.executor, params);
    }

    public void downloadFile(String url, String hash, String fileName, DownloadFileListener listener) {
        DownloadTaskParams params = new DownloadTaskParams();
        params.type = DownloadTaskParams.TASK_TYPE_PLAIN_DOWNLOAD;
        params.url = url;
        params.hash = hash;
        params.listener = listener;
        params.targetFile = new File(rootDir, fileName);
//        params.unzipDirectory = new File(rootDir, hash);
        new DownloadTask(context).executeOnExecutor(this.executor, params);
    }

    public void downloadPatchFromApk(String url, String hash, DownloadFileListener listener) {
        DownloadTaskParams params = new DownloadTaskParams();
        params.type = DownloadTaskParams.TASK_TYPE_PATCH_FROM_APK;
        params.url = url;
        params.hash = hash;
        params.listener = listener;
        params.targetFile = new File(rootDir, hash + ".apk.patch");
        params.unzipDirectory = new File(rootDir, hash);
        new DownloadTask(context).executeOnExecutor(this.executor, params);
    }

    public void downloadPatchFromPpk(String url, String hash, String originHash, DownloadFileListener listener) {
        DownloadTaskParams params = new DownloadTaskParams();
        params.type = DownloadTaskParams.TASK_TYPE_PATCH_FROM_PPK;
        params.url = url;
        params.hash = hash;
        params.originHash = originHash;
        params.listener = listener;
        params.targetFile = new File(rootDir, originHash + "-" + hash + ".ppk.patch");
        params.unzipDirectory = new File(rootDir, hash);
        params.originDirectory = new File(rootDir, originHash);
        new DownloadTask(context).executeOnExecutor(this.executor, params);
    }

    private SharedPreferences sp;

    public void switchVersion(String hash) {
        if (!new File(rootDir, hash+"/index.bundlejs").exists()) {
            throw new Error("Bundle version " + hash + " not found.");
        }
        String lastVersion = getCurrentVersion();
        SharedPreferences.Editor editor = sp.edit();
        editor.putString("currentVersion", hash);
        if (lastVersion != null) {
            editor.putString("lastVersion", lastVersion);
        }
        editor.putBoolean("firstTime", true);
        editor.putBoolean("firstTimeOk", false);
        editor.putBoolean("rolledBack", false);
        editor.apply();
    }

    public void setUuid(String uuid) {
        SharedPreferences.Editor editor = sp.edit();
        editor.putString("uuid", uuid);
        editor.apply();
    }

    public void setBlockUpdate(int until, String reason) {
        SharedPreferences.Editor editor = sp.edit();
        editor.putInt("blockUntil", until);
        editor.putString("blockReason", reason);
        editor.apply();
    }

    public String getCurrentVersion() {
        return sp.getString("currentVersion", null);
    }

    public boolean isFirstTime() {
        return sp.getBoolean("firstTime", false);
    }

    public boolean isRolledBack() {
        return sp.getBoolean("rolledBack", false);
    }

    public void markSuccess() {
        SharedPreferences.Editor editor = sp.edit();
        editor.putBoolean("firstTimeOk", true);
        editor.remove("lastVersion");
        editor.apply();

        this.cleanUp();
    }

    public void clearFirstTime() {
        SharedPreferences.Editor editor = sp.edit();
        editor.putBoolean("firstTime", false);
        editor.apply();

        this.cleanUp();
    }

    public void clearRollbackMark() {
        SharedPreferences.Editor editor = sp.edit();
        editor.putBoolean("rolledBack", false);
        editor.apply();

        this.cleanUp();
    }


    public static void setCustomInstanceManager(ReactInstanceManager instanceManager) {
        mReactInstanceManager = instanceManager;
    }

    public ReactInstanceManager getCustomReactInstanceManager() {
        return mReactInstanceManager;
    }

    public static String getBundleUrl(Context context) {
        return new UpdateContext(context.getApplicationContext()).getBundleUrl();
    }

    public static String getBundleUrl(Context context, String defaultAssetsUrl) {
        return new UpdateContext(context.getApplicationContext()).getBundleUrl(defaultAssetsUrl);
    }

    public String getBundleUrl() {
        return this.getBundleUrl((String) null);
    }

    public String getBundleUrl(String defaultAssetsUrl) {
        isUsingBundleUrl = true;
        String currentVersion = getCurrentVersion();
        if (currentVersion == null) {
            return defaultAssetsUrl;
        }
        // Test should rollback.
        if (!sp.getBoolean("firstTime", false)) {
            if (!sp.getBoolean("firstTimeOk", true)) {
                // Not firstTime, but not ok, so we roll back.
                currentVersion = this.rollBack();
            }
        }

        while (currentVersion != null) {
            File bundleFile = new File(rootDir, currentVersion+"/index.bundlejs");
            if (!bundleFile.exists()) {
                Log.e("getBundleUrl", "Bundle version " + currentVersion + " not found.");
                currentVersion = this.rollBack();
                continue;
            }
            return bundleFile.toString();
        }

        return defaultAssetsUrl;
    }

    private String rollBack() {
        String lastVersion = sp.getString("lastVersion", null);
        SharedPreferences.Editor editor = sp.edit();
        if (lastVersion == null) {
            editor.remove("currentVersion");
        } else {
            editor.putString("currentVersion", lastVersion);
        }
        editor.putBoolean("firstTimeOk", true);
        editor.putBoolean("firstTime", false);
        editor.putBoolean("rolledBack", true);
        editor.apply();
        return lastVersion;
    }

    private void cleanUp() {
        DownloadTaskParams params = new DownloadTaskParams();
        params.type = DownloadTaskParams.TASK_TYPE_CLEANUP;
        params.hash = sp.getString("currentVersion", null);
        params.originHash = sp.getString("lastVersion", null);
        params.unzipDirectory = rootDir;
        new DownloadTask(context).executeOnExecutor(this.executor, params);
    }
}

File: harmony/src/main/ets/UpdateContext.ts
import preferences from '@ohos.data.preferences';
import bundleManager from '@ohos.bundle.bundleManager';
import fileIo from '@ohos.file.fs';
import { DownloadTask } from './DownloadTask';
import common from '@ohos.app.ability.common';
import { DownloadTaskParams } from './DownloadTaskParams';

export class UpdateContext {
    private context: common.UIAbilityContext;
    private rootDir: string;
    private preferences: preferences.Preferences;
    private static DEBUG: boolean = false;
    private static isUsingBundleUrl: boolean = false;

    constructor(context: common.UIAbilityContext) {
        this.context = context;
        this.rootDir = context.filesDir + '/_update';
        
        try {
            if (!fileIo.accessSync(this.rootDir)) {
                fileIo.mkdirSync(this.rootDir);
            }
        } catch (e) {
            console.error('Failed to create root directory:', e);
        }
        this.initPreferences();
    }

    private initPreferences() {
        try {
            this.preferences = preferences.getPreferencesSync(this.context, {name:'update'});
            const packageVersion =  this.getPackageVersion();
            const storedVersion =  this.preferences.getSync('packageVersion', '');
            if(!storedVersion){
                this.preferences.putSync('packageVersion', packageVersion);
                this.preferences.flush();
            } else if (storedVersion && packageVersion !== storedVersion) {
                this.preferences.clear();
                this.preferences.putSync('packageVersion', packageVersion);
                this.preferences.flush();
                this.cleanUp();
            }
        } catch (e) {
            console.error('Failed to init preferences:', e);
        }
    }

    public setKv(key: string, value: string): void {
         this.preferences.putSync(key, value);
         this.preferences.flush();
    }

    public  getKv(key: string): string {
        return  this.preferences.getSync(key, '') as string;
    }

    public  isFirstTime(): boolean {
        return  this.preferences.getSync('firstTime', false) as boolean;
    }

    public  rolledBackVersion(): string {
        return  this.preferences.getSync('rolledBackVersion', '') as string;
    }

    public markSuccess(): void {
        this.preferences.putSync('firstTimeOk', true);
        const lastVersion = this.preferences.getSync('lastVersion', '') as string;
        const curVersion =  this.preferences.getSync('currentVersion', '') as string;
        
        if (lastVersion && lastVersion !== curVersion) {
             this.preferences.deleteSync('lastVersion');
             this.preferences.deleteSync(`hash_${lastVersion}`);
        }
        this.preferences.flush();
        this.cleanUp();
    }

    public clearFirstTime(): void {
        this.preferences.putSync('firstTime', false);
        this.preferences.flush();
        this.cleanUp();
    }

    public clearRollbackMark(): void {
        this.preferences.putSync('rolledBackVersion', null);
        this.preferences.flush();
        this.cleanUp();
    }

    public async downloadFullUpdate(url: string, hash: string, listener: DownloadFileListener): Promise<void> {
        try {
            const params = new DownloadTaskParams();
            params.type = DownloadTaskParams.TASK_TYPE_PATCH_FULL;
            params.url = url;
            params.hash = hash;
            params.listener = listener;
            params.targetFile = `${this.rootDir}/${hash}.ppk`;
            const downloadTask = new DownloadTask(this.context);
            await downloadTask.execute(params);
        } catch (e) {
            console.error('Failed to download full update:', e);
        }
    }

    public async downloadFile(url: string, hash: string, fileName: string, listener: DownloadFileListener): Promise<void> {
        const params = new DownloadTaskParams();
        params.type = DownloadTaskParams.TASK_TYPE_PLAIN_DOWNLOAD;
        params.url = url;
        params.hash = hash;
        params.listener = listener;
        params.targetFile = this.rootDir + '/' + fileName;

        const downloadTask = new DownloadTask(this.context);
        await downloadTask.execute(params);
    }

    public async downloadPatchFromPpk(url: string, hash: string, originHash: string, listener: DownloadFileListener): Promise<void> {
        const params = new DownloadTaskParams();
        params.type = DownloadTaskParams.TASK_TYPE_PATCH_FROM_PPK;
        params.url = url;
        params.hash = hash;
        params.originHash = originHash;
        params.listener = listener;
        params.targetFile = `${this.rootDir}/${originHash}_${hash}.ppk.patch`;
        params.unzipDirectory = `${this.rootDir}/${hash}`;
        params.originDirectory = `${this.rootDir}/${params.originHash}`;
        
        const downloadTask = new DownloadTask(this.context);
        await downloadTask.execute(params);
    }

    public async downloadPatchFromPackage(url: string, hash: string, listener: DownloadFileListener): Promise<void> {
        try {
            const params = new DownloadTaskParams();
            params.type = DownloadTaskParams.TASK_TYPE_PATCH_FROM_APP;
            params.url = url;
            params.hash = hash;
            params.listener = listener;
            params.targetFile = `${this.rootDir}/${hash}.app.patch`;
            params.unzipDirectory = `${this.rootDir}/${hash}`;

            const downloadTask = new DownloadTask(this.context);
            return  await downloadTask.execute(params);
        } catch (e) {
            throw e;
            console.error('Failed to download APK patch:', e);
        }
    }

    public switchVersion(hash: string): void {
        try {
            const bundlePath = `${this.rootDir}/${hash}/bundle.harmony.js`;
            if (!fileIo.accessSync(bundlePath)) {
                throw new Error(`Bundle version ${hash} not found.`);
            }

            const lastVersion = this.getKv('currentVersion');
            this.setKv('currentVersion', hash);
            if (lastVersion && lastVersion !== hash) {
                this.setKv('lastVersion', lastVersion);
            }

             this.setKv('firstTime', 'true');
             this.setKv('firstTimeOk', 'false');
             this.setKv('rolledBackVersion', "");
        } catch (e) {
            console.error('Failed to switch version:', e);
        }
    }

    public static getBundleUrl(context: common.UIAbilityContext, defaultAssetsUrl?: string): string {
        return new UpdateContext(context).getBundleUrl(defaultAssetsUrl);
    }

    public getBundleUrl(defaultAssetsUrl?: string): string {
        UpdateContext.isUsingBundleUrl = true;
        const currentVersion = this.getCurrentVersion();
        if (!currentVersion) {
            return defaultAssetsUrl;
        }
        if (!this.isFirstTime()) {
            if (!this.preferences.getSync('firstTimeOk', true)) {
                return this.rollBack();
            }
        }
        let version = currentVersion;
        while (version) {
            const bundleFile = `${this.rootDir}/${version}/bundle.harmony.js`;
            try {
                if (!fileIo.accessSync(bundleFile)) {
                    console.error(`Bundle version ${version} not found.`);
                    version = this.rollBack();
                    continue;
                }
                return bundleFile;
            } catch (e) {
                console.error('Failed to access bundle file:', e);
                version = this.rollBack();
            }
        }
        return defaultAssetsUrl;
    }

    getPackageVersion(): string {
        let bundleFlags = bundleManager.BundleFlag.GET_BUNDLE_INFO_WITH_REQUESTED_PERMISSION;
        let packageVersion = '';
        try {
            const bundleInfo = bundleManager.getBundleInfoForSelfSync(bundleFlags);
            packageVersion = bundleInfo?.versionName || "Unknown";
        } catch (error) {
            console.error("获取包信息失败:", error);
        }
        return packageVersion;
    }

    public getCurrentVersion() : string {
        const currentVersion = this.getKv('currentVersion');
        return currentVersion;
    }

    private rollBack(): string {
        const lastVersion = this.preferences.getSync('lastVersion', '') as string;
        const currentVersion = this.preferences.getSync('currentVersion', '') as string;
        if (!lastVersion) {
            this.preferences.deleteSync('currentVersion');
        } else {
            this.preferences.putSync('currentVersion', lastVersion);
        }
         this.preferences.putSync('firstTimeOk', true);
         this.preferences.putSync('firstTime', false);
         this.preferences.putSync('rolledBackVersion', currentVersion);
         this.preferences.flush();
        return lastVersion;
    }

    private cleanUp(): void {
        const params = new DownloadTaskParams();
        params.type = DownloadTaskParams.TASK_TYPE_CLEANUP;
        params.hash =  this.preferences.getSync('currentVersion', '') as string;
        params.originHash =  this.preferences.getSync('lastVersion', '') as string;
        params.unzipDirectory = this.rootDir;
        const downloadTask = new DownloadTask(this.context);
        downloadTask.execute(params);
    }

    public getIsUsingBundleUrl(): boolean {
        return UpdateContext.isUsingBundleUrl;
    }
}

export interface DownloadFileListener {
    onDownloadCompleted(params: DownloadTaskParams): void;
    onDownloadFailed(error: Error): void;
}